File: frontend/src/constants/mealTypes.ts
// Mirror of backend/app/core/meal_types.py — keep the two files in sync by hand.
// The backend's MealType enum is the authority; this file exists so the frontend
// can render dropdowns and labels without round-tripping through the API.

export const MEAL_TYPES = [
  "sweet_breakfast",
  "savory_breakfast",
  "brunch",
  "snack",
  "soup",
  "light_lunch",
  "main_course",
  "side_dish",
  "hot_dinner",
  "cold_dinner",
  "dessert",
] as const;

export type MealType = (typeof MEAL_TYPES)[number];

export const MEAL_TYPE_LABELS: Record<MealType, string> = {
  sweet_breakfast: "Sweet breakfast",
  savory_breakfast: "Savory breakfast",
  brunch: "Brunch",
  snack: "Snack",
  soup: "Soup",
  light_lunch: "Light lunch",
  main_course: "Main course",
  side_dish: "Side dish",
  hot_dinner: "Hot dinner",
  cold_dinner: "Cold dinner",
  dessert: "Dessert",
};

const LEGACY_MEAL_TYPE_LABELS: Record<string, string> = {
  breakfast: "Breakfast",
  lunch: "Lunch",
  dinner: "Dinner",
};

// Resolve a label for any meal_type string. Prefer the server-provided localized
// label when present; fall back to the English enum label; last resort is a
// titlecased version of whatever the string is (covers legacy rows).
export function mealTypeLabel(mealType: string, serverLabel?: string | null): string {
  if (serverLabel && serverLabel.trim()) return serverLabel;
  if (mealType in MEAL_TYPE_LABELS) return MEAL_TYPE_LABELS[mealType as MealType];
  if (mealType in LEGACY_MEAL_TYPE_LABELS) return LEGACY_MEAL_TYPE_LABELS[mealType];
  return mealType.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
}

File: backend/app/core/meal_types.py
"""Single source of truth for the user-facing meal-slot taxonomy.

Mirror: ``frontend/src/constants/mealTypes.ts`` — keep the two files in sync by hand.
"""

from __future__ import annotations

from enum import StrEnum


class MealType(StrEnum):
    # StrEnum (Python 3.11+) makes ``str(MealType.SNACK) == "snack"`` and lets
    # SQLAlchemy/JSON serialize instances to their raw values without having to
    # touch ``.value`` at every call site.
    SWEET_BREAKFAST = "sweet_breakfast"
    SAVORY_BREAKFAST = "savory_breakfast"
    BRUNCH = "brunch"
    SNACK = "snack"
    SOUP = "soup"
    LIGHT_LUNCH = "light_lunch"
    MAIN_COURSE = "main_course"
    SIDE_DISH = "side_dish"
    HOT_DINNER = "hot_dinner"
    COLD_DINNER = "cold_dinner"
    DESSERT = "dessert"


# English display strings. The LLM produces a localized ``meal_type_label`` on
# ``PlannedMeal`` for non-English users; these are the backend-side fallbacks
# and the values the frontend shows when no localized label exists.
MEAL_TYPE_LABELS: dict[MealType, str] = {
    MealType.SWEET_BREAKFAST: "Sweet breakfast",
    MealType.SAVORY_BREAKFAST: "Savory breakfast",
    MealType.BRUNCH: "Brunch",
    MealType.SNACK: "Snack",
    MealType.SOUP: "Soup",
    MealType.LIGHT_LUNCH: "Light lunch",
    MealType.MAIN_COURSE: "Main course",
    MealType.SIDE_DISH: "Side dish",
    MealType.HOT_DINNER: "Hot dinner",
    MealType.COLD_DINNER: "Cold dinner",
    MealType.DESSERT: "Dessert",
}


# Map the four legacy values (breakfast/lunch/dinner/snack) onto the new
# taxonomy so existing ``MealEntry.meal_json`` rows still deserialize as
# ``PlannedMeal``. Applied via a pre-validator on ``PlannedMeal.meal_type``.
#
# ``breakfast`` is genuinely ambiguous (could be sweet or savory) — we pick
# ``savory_breakfast`` as the less-lossy default because historical rows skew
# savory and re-labelling a pancake "savory_breakfast" is a smaller UX error
# than re-labelling an omelette "sweet_breakfast". A follow-up backfill can
# refine individual rows using meal name/ingredient heuristics if needed.
LEGACY_MEAL_TYPE_MAP: dict[str, MealType] = {
    "breakfast": MealType.SAVORY_BREAKFAST,
    "lunch": MealType.LIGHT_LUNCH,
    "dinner": MealType.HOT_DINNER,
    "snack": MealType.SNACK,
}
